File: ios/Adapters/LivenessAdapter.swift
import Foundation

/// Common contract for provider adapters on iOS.
///
/// All adapters take a bootstrap JSON object (server-supplied) and
/// resolve with a JSON string describing the completed session, or
/// throw a code/message pair the bridge surfaces as a JS Error.
protocol LivenessAdapter {
  static var providerName: String { get }
  func run(bootstrap: [String: Any], timeoutMs: Int?, locale: String?) async throws -> String
}

struct LivenessAdapterRegistry {
  /// Resolve an adapter by provider name. Returns nil when the
  /// underlying SDK was not compiled in.
  static func adapter(forProviderName name: String) -> LivenessAdapter? {
    switch name {
    case RekognitionLivenessAdapter.providerName:
      return RekognitionLivenessAdapter()
    case IProovLivenessAdapter.providerName:
      return IProovLivenessAdapter()
    default:
      return nil
    }
  }
}

File: android/src/main/java/expo/modules/passkeyliveness/adapters/LivenessAdapter.kt
package expo.modules.passkeyliveness.adapters

import android.app.Activity
import org.json.JSONObject

interface LivenessAdapter {
  val providerName: String

  /**
   * Run the PAD ceremony.
   *
   * @return JSON-stringified completion payload (sessionId, etc.)
   * @throws expo.modules.passkeyliveness.LivenessException on cancel,
   *         permission denial, provider failure, or missing SDK.
   */
  suspend fun run(
    activity: Activity,
    bootstrap: JSONObject,
    timeoutMs: Int?,
    locale: String?,
  ): String
}

object LivenessAdapterRegistry {
  fun adapter(forProviderName: String): LivenessAdapter? = when (forProviderName) {
    RekognitionLivenessAdapter.PROVIDER_NAME -> RekognitionLivenessAdapter()
    IProovLivenessAdapter.PROVIDER_NAME -> IProovLivenessAdapter()
    else -> null
  }
}
